src/gravel/controllers/kv.py

```diff
@@ -12,125 +12,395 @@
-                await asyncio.sleep(1.0)
-                continue
-            opened = True
-            break
-        if opened:
-            self._client = aetcd3.client()
-            self._is_open = True
-            logger.info("opened kvstore connection")
+                if not self._cluster:
+                    # uses /etc/ceph/ceph.client.admin.keyring
+                    # really should do separate keys per node so they can be
+                    # evicted if necessary if nodes are decommisioned
+                    self._cluster = rados.Rados(conffile="/etc/ceph/ceph.conf")
+                logger.debug(f"cluster state: {self._cluster.state}")
+                if self._cluster.state != "connected":
+                    # this can throw (auth failed, etc.)
+                    logger.debug("Connecting to cluster")
+                    self._cluster.connect()
+                    has_aquarium_pool = "aquarium" in self._cluster.list_pools()
+                    if not has_aquarium_pool:
+                        logger.info("Creating aquarium pool")
+                        # consider setting pg_num 1 as with device_health_metrics pool
+                        # also need to do the equivalent of
+                        # `ceph osd pool application enable aquarium aquarium`
+                        # (see devicehealth mgr module)
+                        self._cluster.create_pool("aquarium")
+                    self._ioctx = self._cluster.open_ioctx("aquarium")
+                    # This actually seems to be safe (doesn't trash existing omap
+                    # data if present, which is neat)
+                    self._ioctx.write_full(
+                        "kvstore",
+                        "# aquarium kv store is in this object's omap\n".encode(
+                            "utf-8"
+                        ),
+                    )
+                    # At this point, iff it's a new pool, new object, etc.
+                    # we need to push everything from our local cache to
+                    # the omap on our kvstore, to populate it with whatever
+                    # may have been set pre-bootstrap.
+                    # Probably want to lock the object here too, in case we get two
+                    # simultaenous starts (and thus omap population)
+                    keys = self._db.keys()
+                    values = list(self._db[k] for k in keys)
+                    if keys and not has_aquarium_pool:
+                        with rados.WriteOpCtx() as op:
+                            logger.info(
+                                f"Pushing {keys} to newly created aquarium pool"
+                            )
+                            self._ioctx.set_omap(op, keys, values)  # type: ignore
+                            self._ioctx.operate_write_op(op, "kvstore")
+                    # TODO: need timeouts, error handlers etc on watch
+                    self._config_watch = self._ioctx.watch(
```

Comment:
(the ioctx object seems to be really lacking any upstream documentation that I can find :-()

Reply:
It's pretty sketchy... I've been bouncing between https://docs.ceph.com/en/latest/rados/api/python/, with the C API as a backup https://docs.ceph.com/en/latest/rados/api/librados/. Also it turns out https://github.com/ceph/ceph/blob/master/src/test/pybind/test_rados.py is a useful source of examples.